171654925 add car crash

knowuh opened the pull request.

kswenson commented on js/components/car-crash-instructions.js:
```diff
@@ -0,0 +1,25 @@
+
+Try starting the car at different locations. What do you notice?
+`
+export default class Instructions extends PureComponent {
```
Seems like a good candidate for a functional component. You're using React 16.13, so hooks are available if you find yourself needing state or lifecycle access.

knowuh replied:
Oops yes! Good point! Copy and paste, guilty as charged.  Thanks for pointing that out Kirk.

knowuh commented:
@kswenson Would you like to review this?  Otherwise I can ask @dstrawberrygirl for one.

Cheers

kswenson commented:
@knowuh I'll take a look.

kswenson commented on js/physics.js:
```diff
@@ -93,3 +94,65 @@ export function calcOutputs ({ initialCarX, gravity, surfaceFriction, rampTopX,
+export function calcMagneticForce (distanceMeters, chargeM1, chargeM2) {
+  const epsilon = 0.00001
+  const effectiveDistance = distanceMeters < epsilon ? epsilon : distanceMeters
+  const permiability = 0.8
```
```suggestion
  const permeability = 0.8
```

kswenson commented on js/physics.js:
```diff
@@ -93,3 +94,65 @@ export function calcOutputs ({ initialCarX, gravity, surfaceFriction, rampTopX,
+  const effectiveDistance = distanceMeters < epsilon ? epsilon : distanceMeters
+  const permiability = 0.8
+  const fallOff = 4 * Math.PI * (effectiveDistance * effectiveDistance)
+  // const fallOff = effectiveDistance * effectiveDistance
```
Can presumably remove this now.

kswenson commented on js/physics.js:
```diff
@@ -93,3 +94,65 @@ export function calcOutputs ({ initialCarX, gravity, surfaceFriction, rampTopX,
+  const permiability = 0.8
+  const fallOff = 4 * Math.PI * (effectiveDistance * effectiveDistance)
+  // const fallOff = effectiveDistance * effectiveDistance
+  const forceNewtons = (permiability * chargeM1 * chargeM2) / fallOff
```
```suggestion
  const forceNewtons = (permeability * chargeM1 * chargeM2) / fallOff
```

kswenson commented on js/physics.js:
```diff
@@ -93,3 +94,65 @@ export function calcOutputs ({ initialCarX, gravity, surfaceFriction, rampTopX,
+export function calcAcceleration (forceN, massKg) {
+  const frictionCoef = 10e-5
+  const friction = massKg * -9.81 * frictionCoef
+  const netForceN = (friction + forceN) > 0 ? friction + forceN : 0
```
```suggestion
  const netForceN = Math.max(0, friction + forceN)
```

kswenson commented on js/physics.js:
```diff
@@ -93,3 +94,65 @@ export function calcOutputs ({ initialCarX, gravity, surfaceFriction, rampTopX,
+// see https://en.wikipedia.org/wiki/Force_between_magnets
+export function calcMagneticForce (distanceMeters, chargeM1, chargeM2) {
+  const epsilon = 0.00001
+  const effectiveDistance = distanceMeters < epsilon ? epsilon : distanceMeters
```
```suggestion
  const effectiveDistance = Math.max(epsilon, distanceMeters)
```

kswenson commented on js/physics.js:
```diff
@@ -93,3 +94,65 @@ export function calcOutputs ({ initialCarX, gravity, surfaceFriction, rampTopX,
+
+  const computeX = (timeNow) => {
+    let deltaT = timeNow - startTime
+    deltaT = deltaT * timeScale
```
```suggestion
    deltaT *= timeScale
```

knowuh commented:
Thanks @kswenson -- these are all great suggestions. I am going to take most of them. 

To get the cars to be a reasonable size one needs to pass in a URL parameter `vehicleHeight=xx` where xx is reasonable value.  In a better world, each type of simulation would have its own set of defaults, so that without passing in URL parameters things would look good.  I might do that refactor Today, although I should probably also just move on :)

kswenson commented:
You're right! It works much better with a URL parameter setting the `vehicleHeight` to a reasonable value. 😉

knowuh commented:
Hello @kswenson I applied your suggestions in 7c63125 .  I also extracted a draggable interface to use in the phone drop and possibly ramp game sims later.

Based on feedback from Dan I changed some physics too in a0e78f3
